### ISSUE TITLE
Restore the training loss for the DeepSeek V4 indexer

### ISSUE THREAD

**Issue by HaoranYu111** [pytorch/torchtitan#4802]
The [TODO in DSV4 attention](https://github.com/pytorch/torchtitan/blob/6c2dadbb3109998092d1164e990c6377b2b8fb50/torchtitan/models/deepseek_v4/attention.py#L55-L57) still says that indexer auxiliary loss was temporarily removed until the general aux-loss mechanism landed. That mechanism is now available through [#3864](https://github.com/pytorch/torchtitan/pull/3864).

On main [`6c2dadb`](https://github.com/pytorch/torchtitan/commit/6c2dadbb3109998092d1164e990c6377b2b8fb50), the indexer produces scores, but `Indexer.select()` only returns integer top-k indices. Those indices do not carry gradients back to the indexer's query/key/weight projections. The main attention loss therefore does not replace the missing auxiliary training path.

A CPU check of the current `Indexer.select()` confirms that a loss on the selected values gives gradients to those values, while `idx_q`, `idx_k` and `idx_w` retain `grad=None`. This is a function-level check on PyTorch 2.12, not a full-model training result. The loss is intentionally absent, so this is a request to restore functionality rather than a newly introduced regression.

Is someone already restoring this alongside [#4452](https://github.com/pytorch/torchtitan/pull/4452)? A useful scope would be:

- Inject the indexer loss through the existing auxiliary-loss mechanism.
- Gather teacher KV by row selection, avoiding the expanded four-dimensional gather intermediate.
- Test teacher detach, repeated/invalid selections, TP/CP normalization and gradients, and activation-checkpoint recomputation.

There is an implementation and local validation from our NPU adaptation, but it still needs migration to the current flattened-token/SPMD layout and representative training validation. [#3800](https://github.com/pytorch/torchtitan/pull/3800) has related indexer work for DeepSeek V3.2; it does not restore the V4 path.

<details>
<summary>CPU check of the top-k gradient path</summary>

```python
import torch
from torchtitan.models.deepseek_v4.compressor import Indexer

torch.manual_seed(42)
idx_q = torch.randn(16, 2, 8, requires_grad=True)
idx_k = torch.randn(4, 8, requires_grad=True)
idx_w = torch.randn(16, 2, requires_grad=True)
indices = Indexer.select(idx_q, idx_k, idx_w, seqlen=16, ratio=4, topk=2)
values = torch.randn(4, 8, requires_grad=True)
values[indices].square().sum().backward()
assert indices.dtype == torch.int64 and not indices.requires_grad
assert all(t.grad is None for t in (idx_q, idx_k, idx_w))
assert values.grad is not None
print([t.grad for t in (idx_q, idx_k, idx_w)])  # [None, None, None]
```

</details>

**Comment by tianyu-l** [comment]
cc @sdmyzlp @Matrix-Z97

**Comment by Matrix-Z97** [comment]
@HaoranYu111 
Thanks for the feedback! We are currently working on this.

We already have a preliminary implementation from our NPU adaptation and are actively migrating it to the current flattened-token / SPMD layout, along with preparing representative training validation.

We plan to open a Draft PR soon so everyone can align on progress and conduct code reviews. If there are any updates, specific code details to discuss, or if you'd like us to run cross-validations on GPU environments, we will keep everyone posted in the PR!
